File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Reaction/ReviewEnrollmentReactionHandler.cs
using CreditCardEnrollment.Common.EventStore;
using CreditCardEnrollment.Common.Reaction;
using CreditCardEnrollment.CreditCard.Enrollment.Aggregate;
using CreditCardEnrollment.CreditCard.Enrollment.Event;
using CreditCardEnrollment.CreditCard.Enrollment.Projection.EnrollmentList;
using static CreditCardEnrollment.Common.Util.IdGenerator;

namespace CreditCardEnrollment.CreditCard.Enrollment.Reaction;

public class ReviewEnrollmentReactionHandler : ReactionHandler {
   private readonly GetEnrollmentList _getEnrollmentList;

   public ReviewEnrollmentReactionHandler(
       PostgresTransactionalEventStore eventStore,
       GetEnrollmentList getEnrollmentList) 
       : base(eventStore) {
       _getEnrollmentList = getEnrollmentList;
   }

   public override void React(Common.Event.Event @event) {
       if (@event is not EnrollmentRequested) {
           return;
       }

       var aggregateAndEventIds = _postgresTransactionalEventStore.FindAggregate(@event.AggregateId);
       var aggregate = aggregateAndEventIds.Aggregate;
       var causationId = aggregateAndEventIds.EventIdOfLastEvent;
       var correlationId = aggregateAndEventIds.CorrelationIdOfLastEvent;

       if (aggregate is not Aggregate.Enrollment enrollment) {
           throw new InvalidOperationException("Aggregate not found");
       }

       if (enrollment.Status != EnrollmentStatus.Requested) {
           return;
       }

       var reactionEventId = GenerateDeterministicId($"ReviewedEnrollment{@event.EventId}");
       if (_postgresTransactionalEventStore.DoesEventAlreadyExist(reactionEventId)) {
           return;
       }

       if (_getEnrollmentList.IsThereAnyAcceptedEnrollmentForUserAndProduct(enrollment.UserId, enrollment.ProductId)) {
           _postgresTransactionalEventStore.SaveEvent(new EnrollmentDeclined {
               EventId = reactionEventId,
               AggregateId = enrollment.AggregateId,
               AggregateVersion = enrollment.AggregateVersion + 1,
               CausationId = causationId,
               CorrelationId = correlationId,
               RecordedOn = DateTime.UtcNow,
               ReasonCode = "ALREADY_ACCEPTED",
               ReasonDescription = "You were already accepted to this product."
           });
           return;
       }

       if (enrollment.AnnualIncomeInCents < 1500000) {
           _postgresTransactionalEventStore.SaveEvent(new EnrollmentDeclined {
               EventId = reactionEventId,
               AggregateId = enrollment.AggregateId,
               AggregateVersion = enrollment.AggregateVersion + 1,
               CausationId = causationId,
               CorrelationId = correlationId,
               RecordedOn = DateTime.UtcNow,
               ReasonCode = "INSUFFICIENT_INCOME",
               ReasonDescription = "Insufficient annual income."
           });
           return;
       }

       _postgresTransactionalEventStore.SaveEvent(new EnrollmentAccepted {
           EventId = reactionEventId,
           AggregateId = enrollment.AggregateId,
           AggregateVersion = enrollment.AggregateVersion + 1,
           CausationId = causationId,
           CorrelationId = correlationId,
           RecordedOn = DateTime.UtcNow,
           ReasonCode = "ALL_CHECKS_PASSED",
           ReasonDescription = "All checks passed."
       });
   }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/creditCard/enrollment/reaction/ReviewEnrollmentReactionHandler.ts
import { ReactionHandler } from '../../../common/reaction/ReactionHandler';
import { EnrollmentRequested } from '../event/EnrollmentRequested';
import { EnrollmentAccepted } from '../event/EnrollmentAccepted';
import { EnrollmentDeclined } from '../event/EnrollmentDeclined';
import { PostgresTransactionalEventStore } from '../../../common/eventStore/PostgresTransactionalEventStore';
import { IdGenerator } from '../../../common/util/IdGenerator';
import {GetEnrollmentList} from "../projection/enrollmentList/GetEnrollmentList";
import {Enrollment} from "../aggregate/Enrollment";
import {Event} from "../../../common/event/Event";
import {inject, injectable} from "tsyringe";

@injectable()
export class ReviewEnrollmentReactionHandler extends ReactionHandler {
    constructor(
        @inject(PostgresTransactionalEventStore)eventStore: PostgresTransactionalEventStore,
        @inject(GetEnrollmentList) private readonly getEnrollmentList: GetEnrollmentList,
    ) {
        super(eventStore);
    }

    async react(event: Event): Promise<void> {
        if (!(event instanceof EnrollmentRequested)) {
            return;
        }

        const aggregateData = await this.postgresTransactionalEventStore.findAggregate<Enrollment>(event.aggregateId);
        const enrollment = aggregateData.aggregate;

        if (enrollment.status !== 'Requested') {
            return;
        }

        const reactionEventId = IdGenerator.generateDeterministicId(`ReviewedEnrollment${event.eventId}`);
        if (await this.postgresTransactionalEventStore.doesEventAlreadyExist(reactionEventId)) {
            return;
        }

        const alreadyAccepted = await this.getEnrollmentList.isThereAnyAcceptedEnrollmentForUserAndProduct(
            enrollment.userId,
            enrollment.productId
        );

        if (alreadyAccepted) {
            const reactionEvent = new EnrollmentDeclined(
                reactionEventId,
                enrollment.aggregateId,
                enrollment.aggregateVersion + 1,
                aggregateData.correlationIdOfLastEvent,
                aggregateData.eventIdOfLastEvent,
                new Date(),
                'ALREADY_ACCEPTED',
                'You were already accepted to this product.'
            );

            await this.postgresTransactionalEventStore.saveEvent(reactionEvent);
            return;
        }

        if (enrollment.annualIncomeInCents < 1500000) {
            const reactionEvent = new EnrollmentDeclined(
                reactionEventId,
                enrollment.aggregateId,
                enrollment.aggregateVersion + 1,
                aggregateData.correlationIdOfLastEvent,
                aggregateData.eventIdOfLastEvent,
                new Date(),
                'INSUFFICIENT_INCOME',
                'Insufficient annual income.'
            );

            await this.postgresTransactionalEventStore.saveEvent(reactionEvent);
            return;
        }

        const reactionEvent = new EnrollmentAccepted(
            reactionEventId,
            enrollment.aggregateId,
            enrollment.aggregateVersion + 1,
            aggregateData.correlationIdOfLastEvent,
            aggregateData.eventIdOfLastEvent,
            new Date(),
            'ALL_CHECKS_PASSED',
            'All checks passed.'
        );

        await this.postgresTransactionalEventStore.saveEvent(reactionEvent);
    }
}
